### PR TITLE
Initial propose

### PR DESCRIPTION
**Pull request by phracek**
[EDITED] Betka, a bot, is used for syncing upstream into downstream.
It is able to create a PR in src.fedoraproject.org in
dist-git repository based on upstream repository

How to review this PR? There are two ways:
1. Running test suite
- Create a file `secrets.env` with values like:
```bash
GIHUB_API_TOKEN=dummy
PAGURE_API_TOKEN=dummy
```
- Run `make test-in-container`

2. Run real sync:
- Get GITHUB_API_TOKEN from `https://github.com/settings/tokens` If it does not exist, then create it
- Get PAGURE_API_TOKEN from `https://src.fedoraproject.org/settings#nav-api-tab`. If it does not exist. then create it.
- Save both tokens into `secrets.env` file
- in a terminal run command `make run` and wait a minute till you don't see messages like:
```bash
betka_1      | [2020-04-02 11:57:55,097: DEBUG/MainProcess] ^-- substep ok
betka_1      | [2020-04-02 11:57:55,097: DEBUG/MainProcess] | Consumer: Starting event loop
betka_1      | [2020-04-02 11:57:55,097: DEBUG/MainProcess] | Worker: Hub.register Pool...
betka_1      | [2020-04-02 11:57:55,098: INFO/MainProcess] celery@docker-desktop ready.
betka_1      | [2020-04-02 11:57:55,098: DEBUG/MainProcess] basic.qos: prefetch_count->4
```
- in a second terminal run a command: `make send-master-sync`.

At the end, a new PR in `https://src.fedoraproject.org/container/s2i-core/pull-requests` should be generated.

For now, we are syncing `https://github.org/sclorg/s2i-base-container` into `src.fedoraproject.org/container/s2i-core` only.

Signed-off-by: Petr "Stone" Hracek <phracek@redhat.com>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    - starts OpenShift cluster locally
+    - adds ServiceAccount and RoleBindings
+- Now start tests by command: `py.test-3 tests/test_openshift_pod.py`
+
+See also [bots-deployment](https://github.com/user-cont/bots-deployment) repo.
+
+## User guide for image maintainers
+
+1. You need to contact us for adding your repository to the configuration of our instance of Betka.
+    - You can directly create PR with proper changes in [betka.yaml](https://github.com/sclorg/betka/blob/master/betka-prod.yaml), see [this section](https://gitlab.cee.redhat.com/platform-eng-core-services/betka#betka-main-configuration-file) for more information.
```
Comment (pkubatrh) [inline README.md]: internal link

File: betka/utils.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+def copy_upstream2downstream(src_parent: Path, dest_parent: Path):
+    """Copies content from upstream repo to downstream repo
+
+     Copies all files/dirs/symlinks from upstream source to dist-git one by one,
+     while removing previous if exists.
+     https://gitlab.cee.redhat.com/platform-eng-core-services/docker-rebuild-helper/blob/master/container_workflow_tool/distgit.py#L265
```
Comment (pkubatrh) [inline betka/utils.py]: internal link

File: examples/python3.yaml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  notifications:
+    email_addresses: [phracek@redhat.com]
+
+upstream-to-downstream:
+  # is betka enabled for this repository
+  # optional - defaults to true
+  enabled: true
+
+  notifications:
+    email_addresses: [rh-software-collections@redhat.com]
```
Comment (pkubatrh) [inline examples/python3.yaml]: internal ML, probably should not keep the address in here. Change to sclorg@redhat.com or remove entirely. Same for the other files.

File: files/home/templates/email_pr_template [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+betka synced the upstream PR {{ template_data.pr_number }} to the downstream PR {{ template_data.downstream_pr }}.
+
+Betka {{ template_data.status }} pull request:
+{{ template_data.pagure }}/{{ template_data.namespace }}/{{ template_data.downstream_repo }}/pull-request/{{ template_data.downstream_pr }}
+
+If you don't want to receive these notifications, please, write to phracek@redhat.com.
+
+For more information about betka, see:
+https://gitlab.cee.redhat.com/platform-eng-core-services/betka/blob/master/README.md
```
Comment (pkubatrh) [inline files/home/templates/email_pr_template]: internal link

File: tests/integration/src/bot-cfg.yml [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+  # Specify if master branch in upstream repository is synced
+  master_checker: true
+  # Should pull requests be synced?
+  pr_checker: false
+  # Either 'upstream_branch_name' or 'upstream_git_path' can be specified
+  # Branch name which is used for sync
+  upstream_branch_name: master
+  # Specify URL to an image used for dist-git source generation. Like
+  image_url: docker-registry.upshift.redhat.com/rhscl/dist-git-generator
```
Comment (pkubatrh) [inline tests/integration/src/bot-cfg.yml]: not sure if this is available externally. We should get a different image with the generator anyway

**Comment by pkubatrh** [comment]
Only did a quick optical check to see if there are some internals leftover. We can work on the rest later. Especially the tests look like they will need a lot of work.

File: README.md [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+- loads main configuration file
+- clones specific downstream repository
+- checks if downstream branches, in dist-git repository, contain specific configuration file `bot-cfg.yml`
+- parses `bot-cfg.yml` file and checks if some of checks like `master_checker` or `pr_checker` are allowed
+
+### master_checker
+- if `master_checker` is enabled, then it checks `master` upstream repository branch for new commits
+    - clones upstream repository in case of new commit and configuration file contains `master_checker` flag
+    - syncs files from upstream into downstream specific branch
+    - fills a pull request into Internal Pagure instance `src.osci.redhat.com` if pull request does not exist yet
```
Comment (pkubatrh) [inline README.md]: internal link

File: examples/python3.yaml (hunk 3 above)
Comment (phracek) [reply examples/python3.yaml]: Remove them, is better

**Comment by phracek** [comment]
https://github.com/user-cont/frambo/pull/19

**Pull request merged**